cmd/utils/cmd.go

```diff
@@ -341,6 +344,99 @@ func ExportPreimages(db ethdb.Database, fn string) error {
 			return err
 		}
 	}
-	log.Info("Exported preimages", "file", fn)
+	log.Info("Exported preimages", "file", fn, "elapsed", common.PrettyDuration(time.Since(start)))
+	return nil
+}
+
+// ImportSnapshot imports a batch of snapshot data into the database
+func ImportSnapshot(db ethdb.Database, fn string) error {
+	log.Info("Importing snapshot data", "file", fn)
+	start := time.Now()
+
+	// Open the file handle and potentially unwrap the gzip stream
+	fh, err := os.Open(fn)
+	if err != nil {
+		return err
+	}
+	defer fh.Close()
+
+	var reader io.Reader = fh
+	if strings.HasSuffix(fn, ".gz") {
+		if reader, err = gzip.NewReader(reader); err != nil {
+			return err
+		}
+	}
+	stream := rlp.NewStream(reader, 0)
+
+	// Import the snapshot in batches to prevent disk trashing
+	batch := db.NewBatch()
+	for {
+		// Read the next entry and ensure it's not junk
+		var blob []byte
+		if err := stream.Decode(&blob); err != nil {
+			if err == io.EOF {
+				break
+			}
+			return err
+		}
+		if len(blob) < common.HashLength+1 {
+			continue
+		}
+		key := blob[:common.HashLength+1]
+		val := blob[common.HashLength+1:]
+		batch.Put(key, val)
```

Comment:
Won't this error on storage data, where the key is even larger? And thus place half the key inside the data-portion?

Reply:
Yes, it's wrong here. The key length of storage snapshot key is 1+2*32. I will change it.

Reply:
Fixed